### ISSUE TITLE
All sources are re-built, even if nothing has changed (using nbsphinx extension)

### ISSUE THREAD

**Issue by mgeier** [sphinx-doc/sphinx#5164]
This problem was introduced in PR #4966. Before that, everything was working fine.

This only happens when the `nbsphinx` extension is used (but maybe other extensions are affected, too?).

#### Procedure to reproduce the problem

This can be reproduced by building the docs for the `nbsphinx` module.
Installation instructions: https://github.com/spatialaudio/nbsphinx/blob/master/CONTRIBUTING.rst

In short:

```
git clone https://github.com/spatialaudio/nbsphinx.git
cd nbsphinx
python3 -m pip install -e . --user
python3 setup.py build_sphinx
```

All source files (including Jupyter notebooks) are built at this point.

If I re-build immediately after that (without touching any files) ...

```
python3 setup.py build_sphinx
```

... all source files are built again.

#### Expected results

The second time the source files should not be processed again.
Instead, there should be this message:

```
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 0 source files that are out of date
updating environment: 0 added, 0 changed, 0 removed
looking for now-outdated files... none found
no targets are out of date.
```

### Reproducible project / your project
- https://github.com/spatialaudio/nbsphinx

### Environment info
- OS: Linux
- Python version: Python 3.6.6rc1+
- Sphinx version: `master` branch, every commit after PR #4966 (commit 3ffde92c54c6e9d594b2a41a365a90dc0dcbac6c)

**Comment by tk0miya** [comment]
Confirmed with this Dockerfile:
```
FROM tk0miya/sphinx-html

RUN apt update; apt install -y pandoc
RUN git clone https://github.com/spatialaudio/nbsphinx.git
WORKDIR /docs/nbsphinx
RUN pip3 install -e . jupyter jupyter_client
RUN pip3 install git+https://github.com/sphinx-doc/sphinx
RUN pip3 install -U setuptools
RUN python3 setup.py build_sphinx
RUN python3 setup.py build_sphinx
```

Sphinx misdetects that some configuration has been changed:
```
Step 9/9 : RUN python3 setup.py build_sphinx
 ---> Running in 2a388558dc96
running build_sphinx
Running Sphinx v1.8.0+
loading pickled environment... done
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 0 source files that are out of date
updating environment: [config changed] 19 added, 1 changed, 0 removed
...
```

And Sphinx-1.7 does not do full rebuild. So this is a bug of master branch!

**Comment by tk0miya** [comment]
I just made #5174. And it seems works fine for me. Could you check it if possible?
```
FROM tk0miya/sphinx-html

RUN apt update; apt install -y pandoc
RUN git clone https://github.com/spatialaudio/nbsphinx.git
WORKDIR /docs/nbsphinx
RUN pip3 install -e . jupyter jupyter_client
RUN pip3 install git+https://github.com/tk0miya/sphinx@5164_fix_incremental_build
RUN pip3 install -U setuptools
RUN python3 setup.py build_sphinx
RUN python3 setup.py build_sphinx
```

BTW, I think it would be nice if `_add_notebook_parser()` uses new APIs to register its components like this:
```
    if sphinxversion_info >= (1, 8):
        app.add_source_suffix('.ipynb', 'jupyter_notebook')
        app.add_source_parser(NotebookParser)
    else:
        ...
```

And let's declare the supported format of the parser as an attribute:
```
class NotebookParser(rst.Parser):
    supported = ('jupyter_notebook')
```

I think this will make your extension simpler in future.

**Comment by mgeier** [comment]
Thanks @tk0miya, #5174 works great for me!

I've also created a PR for `nbsphinx` based on your suggestion: https://github.com/spatialaudio/nbsphinx/pull/197

**Comment by tk0miya** [comment]
I just merged #5174. Thank you for confirmation!
